testy.go

```diff
@@ -0,0 +1,133 @@
+package testy
+
+import (
+	"fmt"
+	"runtime"
+	"strings"
+	"sync"
+	"testing"
+
+	"github.com/gametimesf/testy/orderedmap"
+)
+
+type testy struct {
+	tests orderedmap.OrderedMap[string, orderedmap.OrderedMap[string, *Test]]
+}
+
+var instance testy
+
+// TestingT is a subset of testing.T that we have to implement for non-`go test` runs.
+//
+// TODO flesh this out with more useful stuff from testing.T
+type TestingT interface {
+	Fail()
+	FailNow()
+	Fatal(args ...interface{})
+	Fatalf(format string, args ...interface{})
+	Errorf(format string, args ...interface{})
+	Helper()
+	Log(args ...interface{})
+	Logf(format string, args ...interface{})
+	Name() string
+}
+
+type Tester func(t TestingT)
+
+type Test struct {
+	Package string
+	Name    string
+	tester  Tester
+	// TODO better results
+	Msgs   []string
+	Failed bool
+}
+
+var regLock sync.Mutex
+
+func RegisterTest(name string, tester Tester) interface{} {
+	// we only care about our immediate caller
+	callers := make([]uintptr, 1)
+	// skip over Callers and ourselves
+	n := runtime.Callers(2, callers[:])
+	pkg := "<unknown>"
+	// make sure it was able to actually get our caller
+	if n > 0 {
+		frames := runtime.CallersFrames(callers)
+		frame, _ := frames.Next()
+		// remove the function name (which is almost certainly "init") and leave just the package name.
+		// as an example, this function is `github.com/gametimesf/testy.RegisterTest`.
+		i := strings.LastIndex(frame.Function, ".")
+		pkg = frame.Function[:i]
+	}
+
+	regLock.Lock()
+	defer regLock.Unlock()
+	if instance.tests == nil {
+		instance.tests = make(orderedmap.OrderedMap[string, orderedmap.OrderedMap[string, *Test]])
```

Comment:
this is a rather annoying, wordy syntax :(

Reply:
Can you define a type for generics like you could do in c++?  `type MyMap orderedmap.OrderedMap[string, orderedmap.OrderedMap[string, *Test]]`

Reply:
I think so, but I don't know if the decreased clarity over what the map actually is would be worth it since this is the only really obtuse line

Reply:
I think a type alias would be worth it for the sake of readability, since this verbose type is quite long, and used in multiple places throughout this file.

Reply:
ok, so I did this, and honestly I think it makes it look less readable because you don't know what you're making, and quick help on the type of the thing (in goland at least) just shows the alias, not what it actually is.

Reply:
see https://github.com/gametimesf/testy/tree/less-readable (which has some relocation refactoring from the branch I've been working on today)

Reply:
we can revisit this in the next PR for further discussion